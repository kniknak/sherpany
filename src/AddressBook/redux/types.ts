export enum Status {
    Initial = 1,
    Success,
    Pending,
    Failure,
}

export type Nationality = "CH" | "ES" | "FR" | "GB"

export interface AddressBookState {
    status: Status
    users: FetchAddressBookUser[][]
    page: number
    nationalities: Set<Nationality>
}

export interface FetchAddressBookUser {
    name: {
        first: string
        last: string
    }
    location: {
        street: {
            name: string
            number: number
        }
        city: string
        state: string
        postcode: string
    }
    email: string
    login: {
        username: string
    }
    phone: string
    cell: string
    id: {
        value: string
    }
    picture: {
        thumbnail: string
    }
}

export interface FetchAddressBookPayload {
    users: FetchAddressBookUser[]
    page: number
}
