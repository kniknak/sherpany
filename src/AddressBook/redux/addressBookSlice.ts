import { createSlice, PayloadAction } from "@reduxjs/toolkit";
import { AddressBookState, FetchAddressBookPayload, Nationality, Status } from "./types";

export const nationalities: Nationality[] = ["CH", "ES", "FR", "GB"];

export const addressBookInitialState: AddressBookState = {
    status: Status.Initial,
    users: [],
    page: 1,
    nationalities: new Set(nationalities),
};

export const addressBookSlice = createSlice({
    name: "counter",
    initialState: addressBookInitialState,
    reducers: {
        request: state => {
            state.status = Status.Pending;
        },
        success: (state, action: { payload: FetchAddressBookPayload }) => {
            state.status = Status.Success;
            state.users[action.payload.page - 1] = action.payload.users;
        },
        failure: state => {
            state.status = Status.Failure;
        },
        toggleNationality: {
            reducer: (state, action: PayloadAction<Nationality>) => {
                if (state.nationalities.has(action.payload)) {
                    state.nationalities.delete(action.payload);
                } else {
                    state.nationalities.add(action.payload);
                }
            },
            prepare: (nationality: Nationality) => ({ payload: nationality }),
        },
    },
});