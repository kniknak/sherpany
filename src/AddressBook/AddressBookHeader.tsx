import React from "react";
import { RootState } from "../App/store";
import Typography from "@material-ui/core/Typography";
import Toolbar from "@material-ui/core/Toolbar";
import AppBar from "@material-ui/core/AppBar";
import MuiTextField from "@material-ui/core/TextField";
import withStyles from "@material-ui/core/styles/withStyles";
import { Status } from "./redux/types";
import { connect, MapStateToProps } from "react-redux";

const TextField = withStyles(theme => ({
    root: {
        width: 100,
        marginLeft: theme.spacing(2),
    },
}))(MuiTextField);

interface StateProps {
    status: Status
}

interface OwnProps {
    status: Status
    search: string
    setSearch: (search: string) => void
}

export type AddressBookHeaderProps = StateProps & OwnProps

export class AddressBookHeader extends React.PureComponent<AddressBookHeaderProps> {
    private onChange: React.ChangeEventHandler<HTMLInputElement> = event => this.props.setSearch(event.target.value)

    render() {
        const { status, search } = this.props

        return (
            <AppBar position="sticky"
                color="default">
                <Toolbar style={{ alignItems: "baseline" }}>
                    <Typography variant="body1">
                        {Status[status]}
                    </Typography>
                    <Typography
                        variant="body1"
                        style={{ flex: 1 }}
                        align="right"
                    >
                        Search:
                    </Typography>
                    <TextField
                        label="Type text"
                        value={search}
                        onChange={this.onChange}
                        margin="normal"
                        variant="outlined"
                        size="small"
                    />
                </Toolbar>
            </AppBar>
        );
    }
}


const mapStateToProps: MapStateToProps<StateProps, {}, Pick<RootState, "addressBook">> = state => ({
    status: state.addressBook.status,
});

export const AddressBookHeaderConnected = connect(mapStateToProps)(AddressBookHeader);

