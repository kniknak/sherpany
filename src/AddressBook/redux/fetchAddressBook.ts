import { createAction } from "redux-api-middleware";
import { addressBookSlice, nationalities } from "./addressBookSlice";
import { FetchAddressBookPayload, FetchAddressBookUser } from "./types";
import { RootState } from "../../App/store";

export const fetchAddressBook = (page: number) => createAction<Pick<RootState, "addressBook">, Promise<FetchAddressBookPayload>, never>({
    endpoint: state => {
        const nat = [...(state.addressBook.nationalities.size
            ? state.addressBook.nationalities
            : nationalities),
        ]
            .join(",");

        return `https://randomuser.me/api/?seed=sheerpany&page=${page}&results=50&nat=${nat}`;
    },
    method: "GET",
    types: [
        addressBookSlice.actions.request.type,
        {
            type: addressBookSlice.actions.success.type,
            payload: (action, state, response) => (
                response
                    .json()
                    .then<FetchAddressBookPayload>((result: FetchAddressBookResponse) => ({
                        users: result.results,
                        page,
                    }))
            ),
        },
        addressBookSlice.actions.failure.type,
    ],
});

interface FetchAddressBookResponse {
    results: FetchAddressBookUser[]
    info: {
        seed: "sheerpany"
        results: number
        page: number
        version: string
    }
}