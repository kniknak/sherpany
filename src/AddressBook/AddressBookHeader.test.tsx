import React from "react";
import { shallow } from "enzyme";
import { AddressBookHeader, AddressBookHeaderProps } from "./AddressBookHeader";
import { Status } from "./redux/types";

describe("AddressBook", () => {
    describe("AddressBookHeader", () => {
        it("renders", () => {
            const props: AddressBookHeaderProps = {
                search: "search",
                setSearch: () => undefined,
                status: Status.Success,
            };

            const subject = shallow(<AddressBookHeader {...props}/>);

            expect(subject).toMatchSnapshot();
        });
    });
});