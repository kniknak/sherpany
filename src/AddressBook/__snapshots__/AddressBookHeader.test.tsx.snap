// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`AddressBook AddressBookHeader renders 1`] = `
<WithStyles(ForwardRef(AppBar))
  color="default"
  position="sticky"
>
  <WithStyles(ForwardRef(Toolbar))
    style={
      Object {
        "alignItems": "baseline",
      }
    }
  >
    <WithStyles(ForwardRef(Typography))
      variant="body1"
    >
      Success
    </WithStyles(ForwardRef(Typography))>
    <WithStyles(ForwardRef(Typography))
      align="right"
      style={
        Object {
          "flex": 1,
        }
      }
      variant="body1"
    >
      Search:
    </WithStyles(ForwardRef(Typography))>
    <WithStyles(WithStyles(ForwardRef(TextField)))
      label="Type text"
      margin="normal"
      onChange={[Function]}
      size="small"
      value="search"
      variant="outlined"
    />
  </WithStyles(ForwardRef(Toolbar))>
</WithStyles(ForwardRef(AppBar))>
`;
